lib/src/main/kotlin/com/swmansion/starknet/data/TypedData.kt

```diff
@@ -276,29 +438,81 @@ data class TypedData private constructor(
-        return value.removeSuffix("*")
+    private fun stripPointer(type: String): String {
+        return type.removeSuffix("*")
+    }
+
+    private fun extractEnumTypes(type: String): List<String> {
+        require(type.isEnum()) { "Type [$type] is not an enum." }
+
+        return type.substring(1, type.length - 1).let {
+            when {
+                it.trim().isEmpty() -> emptyList()
+                else -> it.split(",").map(String::trim)
+            }
+        }
     }
 
     fun getStructHash(typeName: String, data: String): Felt {
         val encodedData = encodeData(typeName, Json.parseToJsonElement(data).jsonObject)
 
-        return StarknetCurve.pedersenOnElements(getTypeHash(typeName), *encodedData.toTypedArray())
+        return hashArray(listOf(getTypeHash(typeName)) + encodedData)
     }
 
     fun getMessageHash(accountAddress: Felt): Felt {
-        return StarknetCurve.pedersenOnElements(
-            Felt.fromShortString("StarkNet Message"),
-            getStructHash("StarkNetDomain", domain),
-            accountAddress,
-            getStructHash(primaryType, message),
+        return hashArray(
+            listOf(
+                Felt.fromShortString("StarkNet Message"),
+                getStructHash(domain.separatorName, Json.encodeToJsonElement(domain).jsonObject),
+                accountAddress,
+                getStructHash(primaryType, message),
+            ),
         )
     }
 
     companion object {
+        private fun getBasicTypes(revision: Revision): Set<String> {
+            return when (revision) {
+                Revision.V0 -> basicTypesV0
+                Revision.V1 -> basicTypesV1
+            }
+        }
+
+        private fun getPresetTypes(revision: Revision): Map<String, List<Type>> {
+            return when (revision) {
+                Revision.V0 -> presetTypesV0
+                Revision.V1 -> presetTypesV1
+            }
+        }
+
+        private val basicTypesV0: Set<String>
+            get() = setOf("felt", "bool", "string", "selector", "merkletree", "raw")
+
+        private val basicTypesV1: Set<String>
+            get() = basicTypesV0 + setOf("enum", "u128", "i128", "ContractAddress", "ClassHash", "timestamp", "shortstring")
+
+        private val presetTypesV0: Map<String, List<Type>>
+            get() = emptyMap()
+
+        private val presetTypesV1: Map<String, List<Type>>
+            get() = mapOf(
+                "u256" to listOf(
+                    StandardType("low", "u128"),
+                    StandardType("high", "u128"),
+                ),
+                "TokenAmount" to listOf(
+                    StandardType("token_address", "ContractAddress"),
+                    StandardType("amount", "u256"),
+                ),
+                "NftId" to listOf(
+                    StandardType("collection_address", "ContractAddress"),
+                    StandardType("token_id", "u256"),
+                ),
+            )
```

Comment:
I was thinking the same tbh, we could really use those. Can be done in this PR but only if made in one-two commits (to make reviewing easier). otherwise, different PR

Reply:
#440